### PR TITLE
Move docker-options functions to functions file, rework phase_file_path

### PR DESCRIPTION
**Pull request by u2mejc**
@josegonzalez clean up per request in PR https://github.com/dokku/dokku/pull/1830

I pulled the docker-options functions out here too, so it won't be part of the first storage plugin commit when I rebase.

### CODE REVIEW

**Comment by michaelshobbs** [comment]
I didn't follow in the other PR what was trying to be accomplished with regards to `phase_file_path`. Can you clarify here?

**Comment by u2mejc** [comment]
docker-options was using a global var $phase_file_path, and it was being set inside nested functions. Both myself and Jose had to dig to find where it was being set. 

Basicly this was the change:
```bash
-    get_phase_file_path $phase
+    local phase_file_path=$(get_phase_file_path $phase)
```

**Comment by michaelshobbs** [comment]
ah! thanks for the clarification

File: plugins/docker-options/functions [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    fi
+  done
+  dokku_log_fail "Phase(s) must be one of [${phases[*]}]"
+}
+
+get_phase_file_path() {
+  local phase=$1
+  [[ $phase ]] || dokku_log_fail "Phase(s) must be one of [${phases[*]}]"
+  local prefix="DOCKER_OPTIONS_"
+  phase_file_path="${DOKKU_ROOT}/${APP}/${prefix}${phase^^}"
```
Comment (josegonzalez) [inline plugins/docker-options/functions]: I think you're still setting this as a global var?

**Comment by michaelshobbs** [comment]
@u2mejc yeah I don't see the diff you referenced either.
```
-    get_phase_file_path $phase
+    local phase_file_path=$(get_phase_file_path $phase)
```

**Comment by u2mejc** [comment]
$ git my bad... Corrected and rebased. :sweat_smile:

File: plugins/docker-options/functions [hunk 2]
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env bash
+[[ " docker-options docker-options:add docker-options:remove help docker-options:help " == *" $1 "* ]] || exit $DOKKU_NOT_IMPLEMENTED_EXIT
+set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
+source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions"
+
+get_app() {
+  [[ -z $1 ]] && dokku_log_fail "Please specify an app to run the command on"
+  verify_app_name "$1"
+  APP="$1"
```
Comment (josegonzalez) [inline plugins/docker-options/functions]: This is still global...

File: plugins/docker-options/functions [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      return 0
+    fi
+  done
+  dokku_log_fail "Phase(s) must be one of [${phases[*]}]"
+}
+
+get_phase_file_path() {
+  local phase=$1
+  local prefix="DOCKER_OPTIONS_"
+  echo "${DOKKU_ROOT}/${APP}/${prefix}${phase^^}"
```
Comment (u2mejc) [inline plugins/docker-options/functions]: This? Returns string from echo.

File: plugins/docker-options/functions [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  local phase=$1
+  local phase_file_path=$2
+  echo "${phase^} options:"
+  sed -e 's/^/    /' $phase_file_path
+}
+
+display_all_phases_options() {
+  local phases=(build deploy run)
+  for phase in "${phases[@]}"; do
+    local phase_file_path=$(get_phase_file_path $phase)
```
Comment (u2mejc) [inline plugins/docker-options/functions]: To set these.

File: plugins/docker-options/functions (hunk 2 above)
Comment (josegonzalez) [inline plugins/docker-options/functions]: Set `APP` as a local var.

**Comment by u2mejc** [comment]
@josegonzalez @michaelshobbs Looks like circle's lint'ing might be a little too confident of itself here, I linted the file before pushing (hence double quoting all of the naked vars). Want to take a look to see if we can merge this?

**Comment by michaelshobbs** [comment]
```
root@dokku:~/dokku# make lint
# these are disabled due to their expansive existence in the codebase. we should clean it up though
# SC2034: VAR appears unused - https://github.com/koalaman/shellcheck/wiki/SC2034
# SC2086: Double quote to prevent globbing and word splitting - https://github.com/koalaman/shellcheck/wiki/SC2086
linting...

In ./plugins/docker-options/commands line 9:
PHASE_FILE_PATH='${DOKKU_ROOT}/${APP}/${PHASE_FILE_PREFIX}${phase^^}'
                ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.

make: *** [lint] Error 123
```

**Comment by u2mejc** [comment]
Thank you @michaelshobbs, did you read my message?

**Comment by michaelshobbs** [comment]
Yes. Fix the lint issue.

**Comment by u2mejc** [comment]
Ah sorry, I wasn't familiar with shellcheck disables, updating...

**Comment by josegonzalez** [comment]
Uh, we don't want to disable it, we want to use double-quotes, no?

**Comment by u2mejc** [comment]
I moved that path to the top of the file since instead of bearing it inside a function so it will be easy to update when we deprecate ~dokku. The script uses eval below.

**Comment by michaelshobbs** [comment]
Given the choice to use `eval`, disabling that particular lint check makes sense. However, we want the `functions` file to be self-sufficient by default. So moving lines 7-9 in the `commands` file to `functions` would make that true.

Additionally, I think that means you can remove the use of eval and subsequently change the single-quotes to double-quotes. Feel free to correct me if I'm wrong there.

**Comment by u2mejc** [comment]
One concern of @josegonzalez, that $APP was being set in some buried functioned, was rectified by setting it in the routine calling the function. 

Related, the path of the file(s) is now editable by simply updating one line at the top instead of multiple locations inside functions and will help us with the deprecation of ~dokku. The eval updates the path based on the APP name, the prefix, and the phase. I can remove the top var and move it back down to use double quotes but it adds more code / complexity to maintain.

**Comment by u2mejc** [comment]
I'm working on a update to move the var to functions and correct the lint errors now.

File: plugins/docker-options/functions [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+display_phase_options() {
+  local phase="$1"
+  local phase_file_path="$2"
+  echo "${phase^} options:"
+  sed -e 's/^/    /' "$phase_file_path"
+}
+
+display_all_phases_options() {
```
Comment (josegonzalez) [inline plugins/docker-options/functions]: This should take in an `APP` as `$1`, and then use that to construct `$PHASE_FILE_PATH`.

File: plugins/docker-options/functions [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  local phases=("${AVAILABLE_PHASES[@]}")
+  for phase in "${phases[@]}"; do
+    local phase_file_path="$(eval echo $PHASE_FILE_PATH)"
+    if [[ -s "$phase_file_path" ]]; then
+       display_phase_options "$phase" "$phase_file_path"
+    fi
+  done
+}
+
+display_passed_phases_options() {
```
Comment (josegonzalez) [inline plugins/docker-options/functions]: Same basically everywhere...

File: plugins/docker-options/functions (hunk 5 above)
Comment (michaelshobbs) [inline plugins/docker-options/functions]: Problem is `PHASE_FILE_PATH` is used in a for loop that defines `phase` which is a component of `PHASE_FILE_PATH`. I think moving towards a more readable version of this plugin would be best.

Comment (josegonzalez) [inline plugins/docker-options/functions]: Fair enough, but in this case `$APP` might not be set before the functions are sourced in, right?

Comment (michaelshobbs) [inline plugins/docker-options/functions]: I think we should bring back `get_phase_file_path()` and call it thusly:

```
...
  for phase in "${passed_phases[@]}"; do
    local phase_file_path=$(get_phase_file_path $phase)
...
```

Comment (josegonzalez) [inline plugins/docker-options/functions]: fine :(

Comment (michaelshobbs) [inline plugins/docker-options/functions]: @josegonzalez Can you enumerate your issue(s) with this suggestion?

Comment (josegonzalez) [inline plugins/docker-options/functions]: I just think that having a global variable makes things more difficult to read, and we should avoid them where possible.

Comment (michaelshobbs) [inline plugins/docker-options/functions]: In the proposed solution, where is there a global variable?

Comment (michaelshobbs) [inline plugins/docker-options/functions]: Oh! I should be more clear here.

```
get_phase_file_path() {
  local phase=$1
  local prefix="DOCKER_OPTIONS_"
  echo "${DOKKU_ROOT}/${APP}/${prefix}${phase^^}"
}
...
  for phase in "${passed_phases[@]}"; do
    local phase_file_path=$(get_phase_file_path $phase)
...
```

Comment (michaelshobbs) [inline plugins/docker-options/functions]: There's no additional global (`DOKKU_ROOT` and `APP` are global all over) in what I'm proposing.

Comment (josegonzalez) [inline plugins/docker-options/functions]: `APP` totally should not be global all over, which is what I'm trying to get away from here as well.

Comment (michaelshobbs) [inline plugins/docker-options/functions]: In the context of a given plugin, `APP` being set globally by a given case in commands is fine. The proposal to add `APP` as an arg to every function seems much too complex for the provided 'safety` you're looking to add.

Comment (josegonzalez) [inline plugins/docker-options/functions]: Okay, I guess as long as tests pass and you are happy, its fine by me.

**Comment by u2mejc** [comment]
Updated to appease the shellcheck gawds. Added get_phase_file_path function per @michaelshobbs.

**Comment by josegonzalez** [comment]
http://www.ingawdswetrust.com/

**Comment by u2mejc** [comment]
Rebased to pick up circleci bash upgrade.

**Comment by josegonzalez** [comment]
I'm pretty sure this broke tests for all the plugins, see [here](https://travis-ci.org/dokku/dokku-redis/builds/103014451). @michaelshobbs and I are investigating, and will probably put out a 0.4.13 release to fix the issue, since 0.4.12 didn't seem to :(

**Comment by michaelshobbs** [comment]
The failed tests are due to the version of bash not matching what is on ubuntu 14.04. See #1862

**Pull request merged**